core/src/test/scala/unit/kafka/controller/ReassignmentManagerTest.scala

```diff
@@ -0,0 +1,224 @@
+  private final val mockPartitionReassignmentHandler = new PartitionReassignmentHandler(null)
+
+  private var partitionReassignmentManager: ReassignmentManager = null
+
+  @Before
+  def setUp(): Unit = {
+    controllerContext = new ControllerContext
+    controllerContext.epoch = controllerEpoch
+    controllerContext.epochZkVersion = zkEpoch
+    mockZkClient = Mockito.mock(classOf[KafkaZkClient])
+    mockTopicDeletionmManager = Mockito.mock(classOf[TopicDeletionManager])
+    mockControllerBrokerRequestBatch = Mockito.mock(classOf[ControllerBrokerRequestBatch])
+    mockReplicaStateMachine = Mockito.mock(classOf[ReplicaStateMachine])
+    mockPartitionStateMachine = Mockito.mock(classOf[PartitionStateMachine])
+    mockBrokerRequestBatch = Mockito.mock(classOf[ControllerBrokerRequestBatch])
+
+    partitionReassignmentManager = new ReassignmentManager(controllerContext, mockTopicDeletionmManager,
+      mockReplicaStateMachine, mockPartitionStateMachine, new StateChangeLogger(0, inControllerContext = true, None),
+      mockZkClient, mockBrokerRequestBatch, mockPartitionReassignmentHandler,
+      controllerEpoch = () => controllerEpoch)
+  }
+
+  /**
+   * Phase A of a partition reassignment denotes the initial trigger of a reassignment.
+   *
+   *  A1. Update ZK with RS = ORS + TRS,
+   *                     AR = TRS - ORS and
+   *                     RR = ORS - TRS.
+   * A2. Update memory with RS = ORS + TRS, AR = TRS - ORS and RR = ORS - TRS
+   * A3. Send LeaderAndIsr request to every replica in ORS + TRS (with the new RS, AR and RR).
+   *     We do this by forcing an update of the leader epoch in zookeeper.
+   * A4. Start new replicas AR by moving replicas in AR to NewReplica state.
+   */
+  @Test
+  def testPhaseAOfPartitionReassignment(): Unit = {
+    /*
+     * Existing assignment is [1,2,3]
+     * We issue a reassignment to [3, 4, 5]
+     */
+    val reassignContext = ReassignedPartitionsContext(Seq(3, 4, 5), null, persistedInZk = false,
+      ongoingReassignmentOpt = None)
+    val expectedAddingReplicas = Seq(4, 5)
+    val expectedFullReplicaSet = Seq(3, 4, 5, 1, 2)
+    val initialAssignment = PartitionReplicaAssignment(Seq(1, 2, 3), Seq(), Seq())
+    val initialLeaderAndIsr = new LeaderAndIsr(1, 1, List(1, 2, 3), zkEpoch)
+    controllerContext.partitionAssignments.put(topic, mutable.Map(tp.partition() -> initialAssignment))
+    controllerContext.updatePartitionFullReplicaAssignment(tp, initialAssignment)
+    controllerContext.partitionsBeingReassigned.put(tp, reassignContext)
+
+    mockAreReplicasInIsr(tp, List(1, 2, 3), initialLeaderAndIsr)
+    val expectedNewAssignment = PartitionReplicaAssignment.fromOldAndNewReplicas(Seq(1, 2, 3), Seq(3, 4, 5))
+    assertEquals(expectedFullReplicaSet, expectedNewAssignment.replicas)
+    // A1. Should update ZK
+    Mockito.doReturn(mockSetDataResponseOK, Nil: _*).when(mockZkClient).setTopicAssignmentRaw(topic, mutable.Map(tp -> expectedNewAssignment), zkEpoch)
```

Comment:
the use of `Nil: _*` as a second argument is due to this being unable to compile otherwise in Scala, due to its auto-tupling feature ([article explaining it](https://medium.com/@mistadikay/when-polymorphism-bites-you-in-the-ass-c54748eb9153)).
There are some workarounds like adding a helper method or using mockito-scala. I opted for repeating the argument for the time being